### ISSUE TITLE
Very large dask arrays

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#3514]
When dealing with very large datasets sometimes the task graph handling becomes a bottleneck.  What are some ways in which this can be addressed medium-to-long term.

As an example consider a petabyte array cut into gigabyte chunks.  This means that each elementwise operation on the array generates a million new tasks.  If each task takes up something like 1ms of overhead, then each operation generates around 16 minutes of overhead on the client and scheduler (assuming distributed for the time being).  If we're doing operations with a hundred or so operations then we're waiting a day or two with just overhead.

How can we address this situation?

1.  **Reduce per-task overhead**:  1ms today is actually fairly pessimistic, we're closer to 200-300us, but this requires some effort on our part to ensure that all tasks are cheaply serializable.  This means no dynamic functions, or anything that strictly requires cloudpickle or generates a nontrivial amount of data in serialized form.
2.  **Fuse tasks**: Currently when we do `2 * x + 1` both the `2 * ` and `+ 1` operations generate a million tasks.  This hurts both when we construct the dask arrays, and when we send them to the scheduler.  Ideally we would pre-fuse operations like these, so that we only ever generated one million tasks that did both the `2 *` and `+ 1`.  This requires that the things we operate on at first to behave differently than dask.arrays as they are currently designed.  We need some sort of higher-level array construct.  This might either be a fully symbolic array (simliar to SymPy) or it might be a modification to dask.array to support [atop fusion](https://github.com/dask/dask/issues/2538).
3.  **Overlap overhead with computation**: the 16 minutes of overhead might not be that big of a deal if the computation itself is large.  In this case it's mostly a UX problem as people feel concerned that their computation hasn't started yet.  If we're able to break apart the graph into reasonable chunks then we might be able to stream the graph to the scheduler in pieces.  Doing this cheaply and reliably is likely to be very hard though.
4.  **Use much larger tasks**: Everything becomes easier if we make our tasks larger.  Memory and compute power continues to grow.

**Comment by martindurant** [comment]
This may be silly, but is there a case for building the graph in local threads?

**Comment by alimanfoo** [comment]
Not sure if this helps at all, but might be worth taking a look at how nextflow handles large task graphs. I saw a talk the other week by nextflow lead developer and he specifically pointed out that they chose not to build the full graph up front because of problems with very large graphs. I don't know exactly what they do instead though.

**Comment by mrocklin** [comment]
> This may be silly, but is there a case for building the graph in local threads?

Currently building the graph is a Pure-Python process, so I don't think that threads would help much here.  Additionally, not all of the costs are in graph construction.  Sending the tasks over the wire and storing them can also be bottlenecks.

**Comment by mrocklin** [comment]
> Not sure if this helps at all, but might be worth taking a look at how nextflow handles large task graphs. I saw a talk the other week by nextflow lead developer and he specifically pointed out that they chose not to build the full graph up front because of problems with very large graphs. I don't know exactly what they do instead though.

Yes, this is a popular technique also with very numerical linear algebra systems.  It's a neat idea and would help with many of these problems.  It would probably require a biggish change to the dask.array codebase.

**Comment by rabernat** [comment]
> In this case it's mostly a UX problem as people feel concerned that their computation hasn't started yet.

Just a comment on the UX problem. If there could be a dask option to output a progress bar while building large graphs, that could help a lot in the interim.

**Comment by tjcrone** [comment]
I've been running into memory issues when generating lists of delayed functions, using delayed, when the list is only a few hundred or maybe 1000 elements in length. This is really surprising. This is on a notebook pod with access to 8 GB of RAM, which to me seems like a moderately large amount. Why would the building of a task graph require so much ram when building lists of delayed functions? What tools are there to debug this issue? Are there better approaches when using delayed that can solve this problem?

**Comment by mrocklin** [comment]
I would like to avoid having this issue used to diagnose particular problems so that it can focus on the bigger picture.  @tjcrone if you're looking to have a conversation about your issue then can I ask you to raise a separate issue?

**Comment by tjcrone** [comment]
Sorry. Sure.

**Comment by jakirkham** [comment]
> ...specifically pointed out that they chose not to build the full graph up front because of problems with very large graphs.

This was my thinking as well when first glancing at this issue. Given there are generators in Python, could imagine one slowly constructing pieces of the graph through this mechanism. There might be a fair bit of trickery here. Not sure how one optimizes such a graph or more directly how one makes good choices about tasks to submit with this kind of graph.

**Comment by rabernat** [comment]
A possibly related issue is that, at least in my use cases, all the tasks first go to just a few workers. Then they get spread around. Here is what my cluster looks like immediately after a big job hits the scheduler. 

![image](https://user-images.githubusercontent.com/1197350/40501879-32580dfe-5f57-11e8-84a2-886a2044549b.png)
(there are 80 workers in this cluster)

This seems like a good way to crash workers.

**Comment by mrocklin** [comment]
That shouldn't happen and is, I think, likely related to something
incidental.  I recommend raising a separate issue about it and having a
conversation there.  To be clear, I'd like to keep this issue about the
general topic of how to handle very large arrays, and not a place for
people to ask for help on current-day problems.

**Comment by rabernat** [comment]
I understand that. I thought this issue of all-tasks-to-one-worker was
indeed a general problem affecting very large arrays, possibly a
low-hanging fruit. Thanks for correcting my misunderstanding.

**Comment by rainwoodman** [comment]
Is it possible to distribute the scheduler or have a hierarchy of schedulers?

**Comment by rainwoodman** [comment]
`base.py` delivers graphs to schedulers.

Individual objects build the dsk graphs at creation time with more or less of a raw API, based on their per-object chunk size settings. 

Perhaps an experiment can start with a special class of objects that only builds super-chunks, and a new type of scheduler / worker / `get`-method (equiv of multiprocess.py or thread.py) that internally create new dsk graphs and run them.

**Comment by crusaderky** [comment]
My personal take, as I very much hit my head for literal months on this problem:

As little as 3 million tasks will make the scheduler implode. In my problem, I had many more.
I personally sponged the problem by introducing many (6000+) intermediate optimization steps. Each step takes the keys from the final array and compresses everything else into it. This works great for embarassignly parallel problems like mine, and sucks for others.

This way, instead of having 8-something million keys to optimize at once, which not only take hours but typically completely kill my 64+ GB of RAM, I end up with a final dask graph of ~300k keys, which the scheduler has no problems with.

See my more in-depth solution and code at #1916.

**Comment by tjb900** [comment]
Regarding the overhead of sending the graph from the Client to the Scheduler, what about instead sending a function which returns the graph (or returns key/value pairs as a generator, perhaps)?  This is essentially a user-controlled form of compression on the graph, and these huge graphs must, I suspect, have a lot of repeating patterns.  Ours certainly do.

Or, rather than a function, perhaps even a more general subgraph-esque object, which could have a method to return the dependencies/dependents for any given key.  This would remove the need to store the graph explicitly on the scheduler side.

**Comment by crusaderky** [comment]
See link above for my take on the issue :)

**Comment by rainwoodman** [comment]
After reading crusaderky's ticket, I think improving around the current single scheduler should suffice for a long time.
